src/blur/meta-blur-actor.c

```diff
@@ -0,0 +1,933 @@
+
+        if (error) {
+            meta_warning ("clearing blur texture data failed: %s\n", error->message);
+        }
+    }
+
+    clutter_stage_ensure_current (clutter_actor_get_stage (self));
+    x = fminf(x, fw);
+    y = fminf(fh - y - height, fh);
+
+    cogl_flush ();
+    meta_bind_texture (priv->gl_target, priv->gl_handle);
+    meta_copy_sub_tex (priv->gl_target, 0, 0, 0, x, y, width, height);
+    meta_bind_texture (priv->gl_target, 0);
+
+#else
+    // slow version, which works....
+    static CoglPixelBuffer *pixbuf = NULL;
+    static CoglBitmap *source_bmp = NULL;
+
+    //FIXME: honor size change
+    
+    if (pixbuf == NULL)
+        pixbuf = cogl_pixel_buffer_new (ctx, width * height * 4, NULL);
+
+    if (source_bmp == NULL)
+        source_bmp = cogl_bitmap_new_from_buffer (pixbuf,
+                COGL_PIXEL_FORMAT_RGBA_8888,
+                width, height,
+                0,
+                0);
+
+#define COGL_READ_PIXELS_NO_FLIP (1L << 30)
+
+    cogl_framebuffer_read_pixels_into_bitmap (cogl_get_draw_framebuffer (),
+            x, y, 
+            COGL_READ_PIXELS_NO_FLIP | COGL_READ_PIXELS_COLOR_BUFFER,
+            source_bmp);
+
+    if (priv->texture != NULL) {
+        cogl_object_unref (priv->texture);
+        priv->texture = NULL;
+    }
+    priv->texture = cogl_texture_2d_new_from_bitmap (source_bmp);
+    cogl_pipeline_set_layer_texture (priv->pipeline, 0, priv->texture);
+    cogl_pipeline_set_layer_filters (priv->pipeline, 0,
+            COGL_PIPELINE_FILTER_LINEAR_MIPMAP_LINEAR,
+            COGL_PIPELINE_FILTER_LINEAR);
+    /*cogl_object_unref (source_bmp);*/
+    /*cogl_object_unref (pixbuf);*/
+#endif
+
+    return TRUE;
+}
+
+
+static void setup_pipeline (MetaBlurActor   *self, cairo_rectangle_int_t *rect)
+{
+    MetaBlurActorPrivate *priv = self->priv;
+
+    if (priv->changed & CHANGED_SIZE) {
+        g_clear_pointer (&priv->texture, cogl_object_unref);
+        priv->changed &= ~CHANGED_SIZE;
+    }
+
+    prepare_texture(self);
+    if (priv->radius && priv->texture) {
+        create_texture (self);
+
+        cogl_pipeline_set_layer_texture (priv->pipeline2, 0, priv->fbTex);
+        cogl_pipeline_set_layer_filters (priv->pipeline2, 0,
+                COGL_PIPELINE_FILTER_LINEAR_MIPMAP_LINEAR,
+                COGL_PIPELINE_FILTER_LINEAR);
+
+        preblur_texture (self);
+    }
+}
+
+static gboolean meta_blur_actor_get_paint_volume (
+        ClutterActor       *actor,
+        ClutterPaintVolume *volume)
+{
+    return clutter_paint_volume_set_from_allocation (volume, actor);
+}
+
+static void
+meta_blur_actor_allocate (ClutterActor        *actor,
+                       const ClutterActorBox  *box,
+                       ClutterAllocationFlags  flags)
+{
+    MetaBlurActor *self = META_BLUR_ACTOR (actor);
+    MetaBlurActorPrivate *priv = self->priv;
+    ClutterActorClass *parent_class;
+
+    if (priv->window) {
```

Comment:
If we have any window attatched to the blur actor, get it's frame rectangle and allocate with this rectangle.